### ISSUE TITLE
XNNPack Error configuring CMake to build an installable package 

### ISSUE THREAD

**Issue by spather** [tensorflow/tensorflow#57658]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Build/Install

### Source

source

### Tensorflow Version

latest master (commit 9906c67e30c2def772d2a9057c98265d904d8b97)

### Custom Code

No

### OS Platform and Distribution

MacOS 12.4

### GCC/Compiler version

Apple clang version 13.1.6 (clang-1316.0.21.2.5)
Target: arm64-apple-darwin21.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

### Current Behaviour?

```shell
When configuring a build to produce an installable package per the instructions (https://www.tensorflow.org/lite/guide/build_cmake#build_installable_package), I get the following error: 


CMake Error: install(EXPORT "tensorflow-liteTargets" ...) includes target "tensorflow-lite" which requires target "XNNPACK" that is not in any export set.
```
```


### Standalone code to reproduce the issue

```shell
This is my command to generate my build files that produces the error I pasted above: 

cmake -GNinja -S ./third_party/tensorflow/tensorflow/lite -B ./deps/build/tensorflow \
  -DTFLITE_ENABLE_INSTALL:BOOL=ON \
  -DCMAKE_FIND_PACKAGE_PREFER_CONFIG:BOOL=ON \
  -DCMAKE_INSTALL_PREFIX:PATH=./deps/install/tensorflow \
  -Dcpuinfo_DIR:PATH=./deps/install/cpuinfo/share/cpuinfo \
  -Druy_DIR:PATH=./deps/install/ruy/lib/cmake/ruy \
  -Dabsl_DIR:PATH=./deps/install/abseil-cpp/lib/cmake/absl \
  -DEigen3_DIR:PATH=./deps/install/eigen/share/eigen3/cmake \
  -DNEON_2_SSE_DIR:PATH=./deps/install/ARM_NEON_2_x86_SSE/lib/cmake/NEON_2_SSE \
  -DFlatbuffers_DIR:PATH=./deps/install/flatbuffers/lib/cmake/flatbuffers
```
All the dependencies (cpuinfo, ruy, absl, eigen, NEON_2_SSE, Flatbuffers) are built from source and installed at the paths specified. 

If I add `-DTFLITE_ENABLE_XNNPACK:BOOL=OFF` then then everything works.
```


### Relevant log output

_No response_</details>

**Comment by Modiug** [comment]
It's not macOS specific.

I am using exactly the same line as given here: https://www.tensorflow.org/lite/guide/build_cmake#build_installable_package (<install_path> replaced ./deps/install) on an Ubuntu 20.04 LTS system and get those messages:

```
CMake Error: install(EXPORT "tensorflow-liteTargets" ...) includes target "tensorflow-lite" which requires target "absl_flags" that is not in any export set.
CMake Error: install(EXPORT "tensorflow-liteTargets" ...) includes target "tensorflow-lite" which requires target "absl_hash" that is not in any export set.
CMake Error: install(EXPORT "tensorflow-liteTargets" ...) includes target "tensorflow-lite" which requires target "absl_status" that is not in any export set.
CMake Error: install(EXPORT "tensorflow-liteTargets" ...) includes target "tensorflow-lite" which requires target "absl_strings" that is not in any export set.
CMake Error: install(EXPORT "tensorflow-liteTargets" ...) includes target "tensorflow-lite" which requires target "absl_synchronization" that is not in any export set.
CMake Error: install(EXPORT "tensorflow-liteTargets" ...) includes target "tensorflow-lite" which requires target "absl_variant" that is not in any export set.
CMake Error: install(EXPORT "tensorflow-liteTargets" ...) includes target "tensorflow-lite" which requires target "ruy" that is not in any export set.
CMake Error: install(EXPORT "tensorflow-liteTargets" ...) includes target "tensorflow-lite" which requires target "XNNPACK" that is not in any export set.
```

**Comment by giuliano-97** [comment]
My understanding is that those libraries should first be installed manually on your system and their install path (or rather, the directory where the cmake configs for that library are located) should be passed during the CMake generation step as described. So I installed each dependencies to /opt (I am running Ubuntu 20.04) in their own install directory and then ran:
```
cmake -S tensorflow/lite -B build -DCMAKE_BUILD_TYPE=RelWithDebInfo \
   -DCMAKE_INSTALL_PREFIX=/opt/tensorflow-lite \
   -DTFLITE_ENABLE_INSTALL=ON \
   -DCMAKE_FIND_PACKAGE_PREFER_CONFIG=ON \
   -Dabsl_DIR=/opt/abseil/lib/cmake/absl \
   -DEigen3_DIR=/usr/lib/cmake/eigen3 -DFlatbuffers_DIR=/opt/flatbuffers/lib/cmake/flatbuffers \
   -DNEON_2_SSE_DIR=/usr/local/lib/cmake/NEON_2_SSE -Dcpuinfo_DIR=/opt/cpuinfo/share/cpuinfo \
   -Druy_DIR=/opt/ruy/lib/cmake/ruy
```
Well, unfortunately that failed since pthreadpool and XNNPACK are not in the export targets for tensorflow-lite:
```
-- Configuring done
CMake Error: install(EXPORT "tensorflow-liteTargets" ...) includes target "tensorflow-lite" which requires target "pthreadpool" that is not in any export set.
CMake Error: install(EXPORT "tensorflow-liteTargets" ...) includes target "tensorflow-lite" which requires target "XNNPACK" that is not in any export set.
```
So I disabled XNNPACK by setting -DTFLITE_ENABLE_XNNPACK=OFF (probably not necessary anyway) and re-run cmake. Still got the same error for pthreadpool. But well, since the build files had already been generated, I tried to run `make install`. That didn't work either since the version of Eigen installed by apt on Ubuntu 20.04 is not compatible. So, I installed the latest version of Eigen from source and passed this instead. It compiled until 40% then the compiler threw the longest template warning I have ever seen - I guess because tflite uses some custom version of Eigen3 or some special set of compile options for it - and then continued to build.  The library was installed and `tensorflow-liteConfig.cmake` was generated.

**Comment by svkeerthy** [comment]
I am facing the same issue. The cmake fails. 

```
cmake ../tensorflow/lite -DTFLITE_ENABLE_INSTALL=ON 
    -DCMAKE_FIND_PACKAGE_PREFER_CONFIG=ON 
    -Dabsl_DIR=<install_path>/install/lib/cmake/absl   
    -DEigen3_DIR=<install_path>/install/share/eigen3/cmake   
    -DFlatbuffers_DIR=<install_path>/install/cmake/flatbuffers   
    -DNEON_2_SSE_DIR=<install_path>/install/lib/cmake/NEON_2_SSE   
    -Dcpuinfo_DIR=<install_path>/install/share/cpuinfo   
    -Druy_DIR=<install_path>/install/lib/cmake/ruy 
    -Dpthreadpool_DIR=<tensorflow_src>/tensorflow/lite/tools/cmake/modules 
    -DXNNPACK_DIR=<tensorflow_src>/tensorflow/lite/tools/cmake/modules
```

with the error:

```
-- Configuring done
CMake Error: install(EXPORT "tensorflow-liteTargets" ...) includes target "tensorflow-lite" which requires target "pthreadpool" that is not in any export set.
CMake Error: install(EXPORT "tensorflow-liteTargets" ...) includes target "tensorflow-lite" which requires target "XNNPACK" that is not in any export set.
-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

I guess there are no `.cmake` files for `XNNPACK` and `pthreadpool`. And, the current [documentation](https://www.tensorflow.org/lite/guide/build_cmake#build_installable_package) does not talk about xnnpack and pthreadpool. 

However the make succeeds by generating `tensorflow-liteConfig.cmake`. But when I try to integrate with another project by using `find_package(tensorflow-lite REQUIRED CONFIG PATHS <path>)` the build fails with the following error:

```
CMake Error at <path>/tensorflow-liteConfig.cmake:51 (include): include could not find requested file: <path>/tensorflow-liteTargets.cmake
```

Any help would be highly appreciated!

**Comment by kutukvpavel** [comment]
I am facing this exact issue on Debian 11.

Btw, there is an effort to package TFLite for Debian, but building a *.deb out of the experimental [repo](https://salsa.debian.org/deeplearning-team/tensorflow/-/tree/experimental/tflite_packaging) took 4 hours (despite Andorid, CUDA etc being disabled by ./configure) only to fail to install, because for some reason TFLite binary was not found in bazel output directory, while clearly being a bazel target. I don't know how is this possible.

**Comment by eesaber** [comment]
## TL;DR
The error 
```
CMake Error: install(EXPORT "tensorflow-liteTargets" ...) includes target "tensorflow-lite" which requires target "XNNPACK" that is not in any export set.
```
results from two repos
- [XNNPACK](https://github.com/google/XNNPACK)
- [pthreadpool](https://github.com/Maratyszcza/pthreadpool) (dependency of XNNPACK)

## How to fix
### XNNPACK
XNNPACK's `CMakeList.txt` failed to integrate with the `FetchContent` build flow. To make it works, modify XNNPACK's `CMakeList.txt`
```diff
- TARGET_INCLUDE_DIRECTORIES(XNNPACK PUBLIC include)
+ TARGET_INCLUDE_DIRECTORIES(XNNPACK
+  PUBLIC
+  "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>"
+  "$<INSTALL_INTERFACE:include>"
+ )

```
and
```diff
- INSTALL(TARGETS XNNPACK
+ INSTALL(TARGETS XNNPACK pthreadpool pthreadpool_interface allocator
+   EXPORT XNNPACK
    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
    PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
+ INSTALL(
+   EXPORT XNNPACK
+   DESTINATION ${CMAKE_INSTALL_LIBDIR}
+ )

```

### pthreadpool
pthreadpool's `CMakeList.txt` failed to integrate with the `FetchContent` build flow. To make it works, modify pthreadpool's `CMakeList.txt` as follow
```diff
- TARGET_INCLUDE_DIRECTORIES(pthreadpool_interface include) 
+ TARGET_INCLUDE_DIRECTORIES(pthreadpool_interface 
+  INTERFACE 
+    "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>"
+    "$<INSTALL_INTERFACE:include>"
+ )
```

**Comment by sachinprasadhs** [comment]
@eesaber , Thanks for investigating on this and providing a solution, would it be possible for you to create a PR for the same in the relevant repositories.
From your observation, I see there is no dependency on TFLite, does this issue still needs to be kept open in this repo?

**Comment by leleliu008** [comment]
If you want to use your own version of xnnpack and pthreadpool, you can do:

```bash
sed -i 's|NOT DEFINED PTHREADPOOL_SOURCE_DIR|FALSE|' CMakeLists.txt                                                                                
sed -i '/find_package(ruy REQUIRED)/a include(./pthreadpool.cmake)' CMakeLists.txt                                                                 
                                                                                                                                                         
cat > pthreadpool.cmake <<EOF                                                                                                                            
add_library(pthreadpool STATIC IMPORTED)                                                                                                                 
set_target_properties(pthreadpool PROPERTIES                                                                                                             
IMPORTED_LOCATION             "${pthreadpool_LIBRARY_DIR}/libpthreadpool.a"                                                                              
INTERFACE_INCLUDE_DIRECTORIES "${pthreadpool_INCLUDE_DIR}"                                                                                               
)                                                                                                                                                        
EOF                                                                                                                                                      
                                                                                                                                                         
cat > tools/cmake/modules/FindXNNPACK.cmake <<EOF                                                                                                        
add_library(xnnpack STATIC IMPORTED)                                                                                                                     
set_target_properties(xnnpack PROPERTIES                                                                                                                 
IMPORTED_LOCATION             "${xnnpack_LIBRARY_DIR}/libxnnpack.a"                                                                                      
INTERFACE_INCLUDE_DIRECTORIES "${xnnpack_INCLUDE_DIR}"                                                                                                   
)                                                                                                                                                        
EOF
```

more details please refer to https://github.com/leleliu008/ndk-pkg-formula-repository-offical-core/blob/master/formula/tensorflow-lite.yml

**Comment by leleliu008** [comment]
![tensorflow-lite-dependencies](https://github.com/tensorflow/tensorflow/assets/3040852/4b6ab49c-4345-4268-b4cb-c6eee26764bc)

https://www.tensorflow.org/lite/guide/build_cmake#specifics_of_android_cross-compilation

document only say how to use system libs of `abseil ruy flatbuffers eigen3 ARM_NEON_2_x86_SSE`, but do not say how to use system libs of `xnnpack` and `pthreadpool`, they prefer to automatically install it when building tensorflow but prebuild library.

**Comment by eesaber** [comment]
> @eesaber , Thanks for investigating on this and providing a solution, would it be possible for you to create a PR for the same in the relevant repositories.
> From your observation, I see there is no dependency on TFLite, does this issue still needs to be kept open in this repo?

@sachinprasadhs
I am willing to create PR if the relevant repositories are accepting them. 

However, even if the PR is accepted and the issue is fixed, we still need to update the `xnnpack.cmake` file in the `tensorflow/lite/tools/cmake/modules/`.
```cmake
# ...
OverridableFetchContent_Declare(
  xnnpack
  GIT_REPOSITORY https://github.com/google/XNNPACK
  # Sync with tensorflow/workspace2.bzl
  GIT_TAG b9d4073a6913891ce9cbd8965c8d506075d2a45a # 👈 to the fixed commit
```
we need to modify the `GIT_TAG` to reference the commit that fixes the issue. This will ensure that the fix is pulled into the codebase when building the project in the future.

So, I recommend keeping this issue open until the `xnnpack.cmake` file is updated to the correct commit

**Comment by spather** [comment]
@eesaber I finally had a chance to try your fix today and I'm having some problems. I'm aware I may not be doing it correctly, so appreciate any guidance. Steps I used:

1. Ran the cmake command from my original post to configure the build and it failed as before (expected). 
2. I went into the build directory where XNNPACK and pthreadpool had been downloaded and made the patches you suggested to `build/tensorflow/xnnpack/CMakeLists.txt` and `build/tensorflow/pthreadpool-source/CMakeLists.txt`.
3. Re-run the cmake command to configure the build. Now I get the following errors:

```
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "cache" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "hardware-config" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "indirection" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "jit" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "logging" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "memory" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "microkernel-utils" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "microparams-init" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "mutex" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "normalization" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "operators" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "operator-run" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "operator-utils" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "packing" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "post-operation" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "microkernels-prod" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "subgraph" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "fxdiv" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "fp16" that is not in any export set.
CMake Error: install(EXPORT "XNNPACK" ...) includes target "pthreadpool" which requires target "fxdiv" that is not in any export set.
-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```

Can you help me figure out what I'm doing wrong? Thanks for your time on this.

**Comment by pkgoogle** [comment]
Hi @spather,

macOS is better supported with bazel than cmake, can you try building with bazel and see if that allows you to continue your work? https://www.tensorflow.org/lite/guide/build_ios

**Comment by joemamaisarealperson** [comment]
> @eesaber I finally had a chance to try your fix today and I'm having some problems. I'm aware I may not be doing it correctly, so appreciate any guidance. Steps I used:
> 
> 1. Ran the cmake command from my original post to configure the build and it failed as before (expected).
> 2. I went into the build directory where XNNPACK and pthreadpool had been downloaded and made the patches you suggested to `build/tensorflow/xnnpack/CMakeLists.txt` and `build/tensorflow/pthreadpool-source/CMakeLists.txt`.
> 3. Re-run the cmake command to configure the build. Now I get the following errors:
> 
> ```
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "cache" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "hardware-config" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "indirection" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "jit" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "logging" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "memory" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "microkernel-utils" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "microparams-init" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "mutex" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "normalization" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "operators" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "operator-run" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "operator-utils" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "packing" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "post-operation" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "microkernels-prod" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "subgraph" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "fxdiv" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "XNNPACK" which requires target "fp16" that is not in any export set.
> CMake Error: install(EXPORT "XNNPACK" ...) includes target "pthreadpool" which requires target "fxdiv" that is not in any export set.
> -- Generating done
> CMake Generate step failed.  Build files cannot be regenerated correctly.
> ```
> 
> Can you help me figure out what I'm doing wrong? Thanks for your time on this.

Hi @spather 
I am running into the exact same issue. Did you managed to solve this?
Thanks!
